src/UserAction/ModalExecutor.php

```diff
@@ -397,15 +397,25 @@ protected function doPreview(View $modal)
      */
     protected function doFinal(View $modal)
     {
-        foreach ($this->actionData['fields'] ?? [] as $field => $value) {
+        foreach ($this->getActionData('fields') as $field => $value) {
             $this->action->getOwner()->set($field, $value);
         }
 
-        $return = $this->action->execute(...$this->_getActionArgs($this->actionData['args'] ?? []));
+        $return = $this->action->execute(...$this->_getActionArgs($this->getActionData('args')));
 
         $this->_jsSequencer($modal, $this->jsGetExecute($return, $this->action->getOwner()->getId()));
     }
 
+    protected function getActionData(string $step): array
+    {
+        return $this->actionData[$step] ?? [];
+    }
+
+    protected function getStep(): string
```

Comment:
bot sure if needed now...

Reply:
When overriding addFormTo - you need to know in which step you are in order to see if you are dealing with 'args' or 'fields'

Reply:
prop can be made protected... But I am ok with both... No more questions for this once.